src/adapter/src/coord/command_handler.rs

```diff
@@ -821,11 +840,42 @@ impl Coordinator {
             // specified to be at `mz_now()` (which is usually the initial refresh)
             // (similarly to how we don't perform refreshes that were specified to be in the
             // past).
+
+            if acquire_read_holds {
+                self.acquire_read_holds_auto_cleanup(
+                    session,
+                    timestamp,
+                    &self.sufficient_collections_all_clusters(resolved_ids.0.iter()),
+                );
+            }
+
+            Ok(Some(timestamp))
         } else {
             Ok(None)
         }
     }
 
+    /// Given ids syntactically appearing in a query ([ResolvedIds]), this function identifies a bundle of storage and
+    /// compute collection ids sufficient for building a dataflow for the query.
+    ///
+    /// As opposed to [crate::coord::indexes::DataflowBuilder::sufficient_collections], this function doesn't need to
+    /// know which cluster the dataflow would be executed on. It simply collects the relevant ids from all clusters.
+    fn sufficient_collections_all_clusters<'a, I>(&self, ids: I) -> CollectionIdBundle
+    where
+        I: IntoIterator<Item = &'a GlobalId>,
+    {
+        let ids = ids.into_iter().collect_vec();
```

Comment:
Low preference thought, ignore if you like: if this has to be a vec, seems reasonable to require the caller pass a vec?

Reply:
Ok, I don't really have a preference, so I've changed it.